### PR TITLE
[Task] #175 es lint and hooks

### PR DESCRIPTION
**Pull request by DougMidgley**
# PR details

## What is the purpose of this pull request? (put an "X" next to an item)

_Please delete options that are not relevant._

- [ ] Documentation update
- [ ] Bug fix
- [ ] New metadata support
- [ ] Enhanced metadata
- [ ] Add a CLI option
- [ ] Add something to the core
- [X] Other, please explain:

## What changes did you make? (Give an overview)

ESLINT validation on PR and Commit Checks

## Is there anything you'd like reviewers to focus on?

Test locally

## Checklist

- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] ESLint & Prettier are not outputting errors or warnings
- [X] README.md updated (if applicable)
- [-] CHANGELOG.md updated
- [-] ran `npm run docs` to update developer docs

### CODE REVIEW

**Comment by JoernBerkefeld** [comment]
could you please update the draft release 3.2.0?
https://github.com/Accenture/sfmc-devtools/releases/edit/untagged-1ed1364d7f9c78e674b9

File: .husky/post-checkout [hunk 1]
```diff
@@ -0,0 +1,7 @@
+#!/bin/sh
+. "$(dirname "$0")/_/husky.sh"
+
+git config commit.template .git/templatemessage
+TICKETID=`git reflog | awk 'NR==1{ print $8; exit }' | LC_ALL=en_US.utf8 grep -oP '\d{3,5}'`
+echo "[POST_CHECKOUT] Setting template commit to #$TICKETID"
+echo "#$TICKETID " > ".git/templatemessage"
```
Comment (JoernBerkefeld) [inline .husky/post-checkout:7]: could you explain what that does? :)

File: package.json [hunk 2]
```diff
@@ -52,8 +53,8 @@
         "sql-formatter-plus": "1.3.6",
         "toposort": "2.0.2",
         "update-notifier-git": "5.0.3",
-        "winston": "3.3.4",
-        "yargs": "17.3.0"
+        "winston": "3.4.0",
+        "yargs": "17.3.1"
```
Comment (JoernBerkefeld) [inline package.json:57]: any reason this was updated? dont see a note in the commits for it. if we don't warn about package changes, developers might not run `npm install` again, leading to issues.

File: package.json [hunk 3]
```diff
@@ -64,7 +65,7 @@
         "eslint-plugin-prettier": "4.0.0",
         "husky": "7.0.4",
         "jsdoc-to-markdown": "7.1.0",
-        "lint-staged": "12.1.5",
+        "lint-staged": "12.1.7",
```
Comment (JoernBerkefeld) [inline package.json:68]: any reason this was updated? dont see a note in the commits for it. if we don't warn about package changes, developers might not run `npm install` again, leading to issues.

File: .husky/commit-msg [hunk 4]
```diff
@@ -0,0 +1,22 @@
+#!/bin/sh
```
Comment (JoernBerkefeld) [inline .husky/commit-msg:1]: again, quick walk-through please

File: package.json [hunk 5]
```diff
@@ -30,7 +30,8 @@
         "build": "run-s lint docs",
         "debug": "node --nolazy --inspect-brk=9229 lib/index.js",
         "docs": "jsdoc2md --files lib/**/*.js > docs/dist/documentation.md",
-        "lint": "eslint --fix lib/**/*.js && eslint --fix test/**/*.js && prettier --write lib/**/*.js",
+        "lint:fix": "eslint --fix lib/**/*.js && eslint --fix test/**/*.js && prettier --write lib/**/*.js",
+        "lint": "eslint lib/**/*.js && eslint test/**/*.js",
```
Comment (JoernBerkefeld) [inline package.json:34]: why the change? for the commit hook?

**Review by JoernBerkefeld — CHANGES_REQUESTED** [review]
changelog needs updating once this is approved

File: package.json (hunk 5 above)
Comment (DougMidgley) [reply package.json:34]: because lint shouldnt fix, it should lint only. This way we can run it in a check and not just hide issues

File: .husky/post-checkout (hunk 1 above)
Comment (DougMidgley) [reply .husky/post-checkout:7]: when you checkeout a branch it takes the branch name and tries to get the ticket id. this then defaults the start of the commit message to #NNNN to help improve compliance in commit messages

File: .husky/commit-msg (hunk 4 above)
Comment (DougMidgley) [reply .husky/commit-msg:1]: basically, this checks the default commit message (other comment) and checks that the commit message matches the correct format. If it is a merge, this is skipped to avoid problems

File: .husky/post-checkout (hunk 1 above)
Comment (JoernBerkefeld) [reply .husky/post-checkout:7]: Fallback?
Is this executed Locally?

File: .husky/commit-msg (hunk 4 above)
Comment (JoernBerkefeld) [reply .husky/commit-msg:1]: Fallback?

Comment (DougMidgley) [reply .husky/commit-msg:1]: meaning?

File: .husky/post-checkout (hunk 1 above)
Comment (DougMidgley) [reply .husky/post-checkout:7]: its executed on any machine, local or not. In a workflow for example it will also execute, unless overriden (which you can also do). What would you look for in a fallback?

Comment (JoernBerkefeld) [reply .husky/post-checkout:7]: when you say "defaults" the commit message - does that impact what people see in their git clients, like sourcetree or gitfork? or where is this "defaulted"?

if it doesn't find an id (by the way, what is it looking for exactly? I cant figure out that grep logic...) it just prints `"# "` into .git/templatemessage?

Comment (JoernBerkefeld) [reply .husky/post-checkout:7]: never mind - im seeing this happens upon stage in gitfork, at the latest upon clicking into the commit message field - unless the field already has content in it

File: .husky/commit-msg (hunk 4 above)
Comment (JoernBerkefeld) [reply .husky/commit-msg:1]: i mean what happens if the default commit message created in `post-checkout` is empty?

Comment (JoernBerkefeld) [reply .husky/commit-msg:1]: also, im seeing a different error message than the one you wrote?
image

File: .husky/post-checkout (hunk 1 above)
Comment (JoernBerkefeld) [reply .husky/post-checkout:7]: tested this some more: if it does not find an id it seems to fail on line 5 and subsequently never update .git/templatemessage again. to test that, I first checked out your branch (175) and played around with the auto-completion. Then, I created a branch off of `develop`, switched to and merged your code in, afterwards I made a change and checked again. It would still autocomplete "#175 ".

same happened when i renamed your 175 branch to something without an id which i tried before the branch off of develop.

- [ ] add fallback for branches that don't have ids in them (e.g. created by external developers in their forked repos)

File: .husky/commit-msg (hunk 4 above)
Comment (JoernBerkefeld) [reply .husky/commit-msg:1]: - [ ] check why the error message would talk about no matching files rather than the wrong commit message (tried committing something without "#123" upfront

**Review by JoernBerkefeld — CHANGES_REQUESTED** [review]

File: .husky/commit-msg (hunk 4 above)
Comment (DougMidgley) [reply .husky/commit-msg:1]: i pushed some changes but couldnt get this error. online search suggests running npm i again

**Review by JoernBerkefeld — APPROVED** [review]
all good now. merging
